mysql/datadog_checks/mysql/execution_plans.py

```diff
@@ -307,32 +312,106 @@ def _collect_execution_plans(self, db, tags, options, min_collection_interval):
+        """
+        Tries the available methods used to explain a statement for the given schema. If a non-retryable
+        error occurs (such as a permissions error), then statements executed under the schema will be
+        disallowed in future attempts.
+        """
+        explain_strategy_none = 'NONE'
+        explain_strategy_procedure = 'PROCEDURE'
+        explain_strategy_statement = 'STATEMENT'
+
+        fns_by_strategy = {
+            explain_strategy_procedure: self._run_explain_procedure,
+            explain_strategy_statement: self._run_explain,
+        }
+
+        plan = None
+        strategy_cache_key = 'explain_strategy:%s' % schema
+        strategy_cache_ttl = 60 * 10
+
+        if not self._can_explain(statement):
+            self.log.debug('Skipping statement which cannot be explained: %s', statement)
+            return None
+
+        if self._expiring_cache.get(strategy_cache_key) == explain_strategy_none:
+            self.log.debug('Skipping statement due to cached collection failure: %s', statement)
+            return None
+
+        exceptions = (pymysql.err.InternalError, pymysql.err.ProgrammingError)
+        non_retryable_errors = frozenset({
+            1044,  # access denied on database
+            1046,  # no permission on statement
+            1049,  # unknown database
+            1305,  # procedure does not exist
+            1370,  # no execute on procedure
+        })
 
         try:
-            start_time = time.time()
-            if schema is not None:
-                cursor.execute('USE `{}`'.format(schema))
-            cursor.execute('EXPLAIN FORMAT=json {statement}'.format(statement=statement))
-            statsd.timing("dd.mysql.run_explain.time", (time.time() - start_time) * 1000, tags=instance_tags)
-        except (pymysql.err.InternalError, pymysql.err.ProgrammingError) as e:
+            # Switch to the right schema; this is necessary when the statement uses non-fully qualified tables
+            # e.g. `select * from mytable` instead of `select * from myschema.mytable`
+            self._use_schema(cursor, schema)
+        except exceptions as e:
             if len(e.args) != 2:
                 raise
-            if e.args[0] in (1046,):
-                self.log.warning('Failed to collect EXPLAIN due to a permissions error: %s, Statement: %s', e.args,
-                                 statement)
-                return None
-            elif e.args[0] == 1064:
-                self.log.warning('Programming error when collecting EXPLAIN: %s, Statement: %s', e.args, statement)
-                return None
-            else:
-                raise
+            if e.args[0] in non_retryable_errors:
+                self._expiring_cache.set(strategy_cache_key, explain_strategy_none, strategy_cache_ttl)
+            self.log.debug('Cannot collect execution plan because %s schema could not be accessed: %s, statement: %s', schema, e.args, statement)
```

Comment:
I think this should be a warning? Seems like something a user would want to see in the logs by default. 

Also could include the full exception message (but not a full stack trace) as that would explain exactly what the problem is.